### ISSUE TITLE
Double value to string conversion difference between netcore and netframework

### ISSUE THREAD

**Issue by shvez** [dotnet/corefx#42576]
I found out that line `Console.WriteLine(123e+20d);` produces different result for netcore and for netframework. 

for value 123e+20d
netcore output is 1.2300000000000001E+22
netframework output is 1.23E+22

Same results for `Convert.ToString(123e+20d, CultureInfo.InvariantCulture)`

Here is sampe app to reproduce the issue:
[ToStringDifference.zip](https://github.com/dotnet/corefx/files/3841841/ToStringDifference.zip)

Is this a bug or expected behavior?
and is there a way to get the same output as for netframework?

**Comment by Joe4evr** [comment]
[This is an intentional change.](https://devblogs.microsoft.com/dotnet/floating-point-parsing-and-formatting-improvements-in-net-core-3-0/)

**Comment by tannergooding** [comment]
Yes, this is an intentional change that is documented as part of the breaking changes: https://docs.microsoft.com/en-us/dotnet/core/compatibility/2.2-3.0#floating-point-formatting-and-parsing-behavior-changed

This particular change was due to double and single-precision strings not being roundtrippable by default. This often lead to confusing scenarios because all three of the following would print as `1.23E+22`:
```csharp
// Value                Bit Representation
1.2299999999999999E+22  0x4484D64651FE74C5
1.2300000000000001E+22  0x4484D64651FE74C6
1.2300000000000003E+22  0x4484D64651FE74C7
```

As such there may be some values that print with additional precision. If you are using them for something like display purposes, you should likely be using [Custom Numeric Format Strings](https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-numeric-format-strings) or explicit [Standard Numeric Format Strings](https://docs.microsoft.com/en-us/dotnet/standard/base-types/standard-numeric-format-strings) (such as `"0.0e+00"` or `"E2"`).

**Comment by tannergooding** [comment]
It is worth noting that in this particular case, we aren't returning the "shortest roundtrippable string", which would be `1.23E+22`. There were some bugs in the `BigInteger` implementation that would cause us to sometimes return too many digits. These were addressed two days ago in https://github.com/dotnet/coreclr/pull/27280.

I'll add a regression test for this particular value before closing the issue.

**Comment by shvez** [comment]
@tannergooding BinInteger? Initially, the report was about double

**Comment by tannergooding** [comment]
Yes, parsing and formatting IEEE 754 floating point values requires utilizing a BigInteger implementation. The one we use is here: https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/Number.BigInteger.cs

**Comment by shvez** [comment]
@tannergooding thank you
